client-tez/org/apache/tez/runtime/library/common/sort/impl/RssSorter.java

```diff
@@ -0,0 +1,196 @@
+        ApplicationAttemptId applicationAttemptId =
+                containerId.getApplicationAttemptId();
+        LOG.info("containerIdStr is {}", containerIdStr);
+        LOG.info("containerId is {}", containerId);
+        LOG.info("applicationAttemptId is {}", applicationAttemptId.toString());
+
+
+        bufferManager = new SortWriteBufferManager(
+                tezTaskAttemptID,
+                (long)(ByteUnit.MiB.toBytes(sortmb) * sortThreshold),
+                applicationAttemptId.toString(),
+                taskAttemptId,
+                successBlockIds,
+                failedBlockIds,
+                RssTezUtils.createShuffleClient(conf),
+                comparator,
+                maxSegmentSize,
+                keySerializer,
+                valSerializer,
+                maxBufferSize,
+                memoryThreshold,
+                sendThreshold,
+                batch,
+                new RssConf(),
+                partitionToServers,
+                numMaps,
+                isMemoryShuffleEnabled(storageType),
+                sendCheckInterval,
+                sendCheckTimeout,
+                bitmapSplitNum,
+                shuffleId);
+        LOG.info("Initialized SortWriteBufferManager.");
+    }
+
+    @Override
+    public void flush() throws IOException {
+        bufferManager.waitSendFinished();
+    }
+
+    @Override
+    public final List<Event> close() throws IOException {
+        super.close();
+        bufferManager.freeAllResources();
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void write(Object key, Object value) throws IOException {
+        try {
+            collect(key, value, partitioner.getPartition(key, value, partitions));
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    synchronized void collect(Object key, Object value, final int partition) throws IOException, InterruptedException {
+        if (key.getClass() != keyClass) {
+            throw new IOException("Type mismatch in key from map: expected "
+                    + keyClass.getName() + ", received "
+                    + key.getClass().getName());
+        }
+        if (value.getClass() != valClass) {
+            throw new IOException("Type mismatch in value from map: expected "
+                    + valClass.getName() + ", received "
+                    + value.getClass().getName());
+        }
+        if (partition < 0 || partition >= partitions) {
+            throw new IOException("Illegal partition for " + key + " ("
+                    + partition + ")");
+        }
+
+        bufferManager.addRecord(partition, key, value);
+        numRecordsPerPartition[partition]++;
+    }
+
+    public int[] getNumRecordsPerPartition() {
+        return numRecordsPerPartition;
+    }
+
+    private boolean isMemoryShuffleEnabled(String storageType) {
```

Comment:
This method should be moved to common package to let every framework can use it.